### ISSUE TITLE
Add support for quantification over tuples

### ISSUE THREAD

**Issue by lemmy** [tlaplus/tlapm#11]
```
File "/home/mkuppe/src/TLA/AsyncGameOfLife/Snippets.tla", line 8, character 18

Operator "x" not found
```

```tla
------------------------------ MODULE Snippets ------------------------------
EXTENDS Integers

works ==
   { <<e[1], e[2]>>: e \in ({1,2} \X {3,4}) }

fails ==
   { <<x, y>>: <<x,y>> \in ({1,2} \X {3,4}) } \* Culprit

THEOREM works <=> fails BY DEF works, fails

=============================================================================
```

Originally discovered in https://github.com/mryndzionek/tlaplus_specs/blob/75c5989af133abc0316414552b729a2af345d1ac/GameOfLife.tla#L25-L28

**Comment by muenchnerkindl** [comment]
TLAPS doesn't support quantification over tuples:

> Quantification over tuples and set constructors using tuples are not supported
> (for example {<<x, y>> \in S \X T : ...}). Note that this produces a really obscure error message, or sometimes a silent crash of TLAPS.

https://tla.msr-inria.inria.fr/tlaps/content/Documentation/Unsupported_features.html

It's been on our to-do list for far too long ...

Stephan

**Comment by lemmy** [comment]
@johnyf Is this also addressed by https://github.com/tlaplus/tlapm/pull/34?

**Comment by johnyf** [comment]
No, #34 did not address tuple declarations in set constructors and constant quantifiers. I now reused some of the functionality from #34 to implement:
- tuple declarations in set constructors for the scheme of set separation
- tuple declarations in set constructors for the scheme of set replacement
- tuple declarations in constant quantifiers
- tuple declarations in `CHOOSE` expressions
- tuple declarations in function definitions (actually this would have been implemented in #34, had I realized that it was not).

In the course of implementing these, I found 3 bugs:
- function definitions allowed declaring unbounded constants (e.g., `f[x] == ...`), which would have been detected by SANY before reaching `tlapm`,
- constant quantifiers allowed mixing unbounded and bounded declarations in the same quantifier (e.g., `\E x \in S, y: ...`), which would have been detected by SANY before reaching `tlapm`,
- tuple declarations in function constructors were incorrectly represented. I introduced this error recently, in #34, by incorrectly implementing the meaning of such declarations. The (correct) definition is in terms of a `CHOOSE` over an unbounded tuple.

Fixing the last bug in this list essentially required implementing tuple declarations for set constructors and (most cases of) `CHOOSE`.

The appearance of `CHOOSE` in the encoding seems to make the proof obligations not automatically provable, at least not without refining the proof to reason about `CHOOSE`. The example given above (https://github.com/tlaplus/tlapm/issues/11#issue-727021829) is now parsed without errors. It does not get proved (actually it did get proved when I first implemented "naively" tuple declarations in set constructors without `CHOOSE`). The output from `tlapm` includes (I changed `<=>` to `=` in the source):

```tla
[ERROR]: Could not prove or check:
           ASSUME works == {<<e[1], e[2]>> : e \in {1, 2} \X {3, 4}},
                  fails ==
                    {LET fcnbnd#x#choose ==
                           CHOOSE fcnbnd#x#choose :
                             \E x, y :
                                /\ fcnbnd#x#choose = <<x, y>>
                                /\ fcnbnd#x = <<x, y>>
                         x == fcnbnd#x#choose[1]
                         y == fcnbnd#x#choose[2]
                     IN  <<x, y>> :
                       fcnbnd#x \in {1, 2} \X {3, 4}}
           PROVE  works = fails
```

This also demonstrates how the encoding looks like, which, in this particular case, is a combination of the definitions on pages 304 and 296 of the book "Specifying Systems".

I also wrote a number of tests. These are tests of parsing, and of failing to parse grammatically incorrect files. I had to comment the only existing theorem-proving test for tuple declarations in function constructors (that was added in #34), because after `CHOOSE` appeared in the proof obligation, that test does not get proved any more. I might try to write a refined proof about what holds for `CHOOSE`.

In particular, I get an error that resembles the errors that I got when trying to write tests for other forms of tuple declarations for function constructors:

```tla
!! Constraint generation. Expression not supported:
[CONSTANT_fcnbnd#m_ \in {1} \X {2} |->
   CHOOSE CONSTANT_fcnbnd#m#choose_ :
     \E CONSTANT_m_, CONSTANT_n_ :
        /\ CONSTANT_fcnbnd#m#choose_ = <<CONSTANT_m_, CONSTANT_n_>>
        /\ CONSTANT_fcnbnd#m_ = <<CONSTANT_m_, CONSTANT_n_>>[1]][1, 2]
Zenon error: exhausted search space without finding a proof
(* created new ".tlacache/parse_function_comma_args.tlaps/parse_function_comma_args.thy" *)
(* fingerprints written in ".tlacache/parse_function_comma_args.tlaps/fingerprints" *)
File "./parse_function_comma_args.tla", line 21, characters 1-2:
[ERROR]: Could not prove or check:
           ASSUME k ==
                    [fcnbnd#m \in {1} \X {2} |->
                       LET fcnbnd#m#choose ==
                             CHOOSE fcnbnd#m#choose :
                               \E m, n :
                                  /\ fcnbnd#m#choose = <<m, n>>
                                  /\ fcnbnd#m = <<m, n>>
                           m == fcnbnd#m#choose[1]
                           n == fcnbnd#m#choose[2]
                       IN  m]
           PROVE  k[1, 2] = 1
File "./parse_function_comma_args.tla", line 1, character 1 to line 33, character 80:
[ERROR]: 1/1 obligation failed.
```

Notice the message "Constraint generation. Expression not supported". It seems that the backends are not prepared to handle these new forms of proof obligations.

In any case, it seems that this issue is more about the parser, than about support in the backends. Support in the backends might be more difficult to develop than the extension of the parser to handle this syntax. At least now the tuple declarations are (hopefully correctly) represented internally, so that further work in the backends can occur.

**Comment by johnyf** [comment]
This implementation of tuple declarations is currently at: https://github.com/johnyf/tlapm/tree/tuple_declarations

**Comment by ahelwer** [comment]
Ref #159

**Comment by ahelwer** [comment]
Fixed by #140

**Comment by ahelwer** [comment]
Actually maybe they were fixed at the parser level but they still need to be translated to the backend provers, @johnyf can you confirm?
